ambry-store/src/main/java/com.github.ambry.store/IndexPersistor.java

```diff
@@ -0,0 +1,167 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.codahale.metrics.Timer;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.utils.Time;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * IndexPersistor to persist index files to disk on a periodic manner.
+ * It also exposes {@link #persist()} to persist on demand.
+ */
+class IndexPersistor implements Runnable {
+
+  private final StoreMetrics metrics;
+  private final ConcurrentSkipListMap<Offset, IndexSegment> indexes;
+  private final Log log;
+  private final String dataDir;
+  private final HardDeleter hardDeleter;
+  private final Time time;
+  private final StoreConfig config;
+  private final Object lock = new Object();
+  private final Object writeLock = new Object();
+  private final AtomicBoolean running;
+  private final CountDownLatch shutdownLatch;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  IndexPersistor(ConcurrentSkipListMap<Offset, IndexSegment> indexes, Log log, String dataDir, Time time,
+      StoreConfig config, HardDeleter hardDeleter, StoreMetrics metrics) {
+    this.indexes = indexes;
+    this.log = log;
+    this.dataDir = dataDir;
+    this.config = config;
+    this.time = time;
+    this.hardDeleter = hardDeleter;
+    this.metrics = metrics;
+    running = new AtomicBoolean(true);
+    shutdownLatch = new CountDownLatch(1);
+  }
+
+  @Override
+  public void run() {
+    try {
+      synchronized (lock) {
+        if (running.get()) {
+          time.wait(lock, (config.storeDataFlushDelaySeconds + new Random().nextInt(Time.SecsPerMin)) * Time.MsPerSec);
+        }
+      }
+      while (running.get()) {
+        synchronized (writeLock) {
+          write();
+        }
+        synchronized (lock) {
+          if (!running.get()) {
+            break;
+          }
+          time.wait(lock, (config.storeDataFlushIntervalSeconds) * Time.MsPerSec);
+        }
+      }
+    } catch (Exception e) {
+      logger.error("Index : " + dataDir + " error while persisting the index to disk ", e);
+    } finally {
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Persists all index files to disk if {@link IndexPersistor} is running
+   * @throws StoreException
+   */
+  void persist() throws StoreException {
+    synchronized (writeLock) {
```

Comment:
From our offline discussion, I thought this was what we wanted to avoid? If I followed our discussion correctly the idea was to let the persistor thread (and only one thread) make the write call. `persist()` is meant to trigger a condition that forces *that* thread to write immediately instead of wait for a certain time. 

If this is our intent (and my understanding is that it is not), then I'm not sure what purpose the condition serves. In this approach `run()`, `persist()` and `shutdown()` can simply contend for one lock and call `write()`. 

Could you clarify?